### DISCUSSION TITLE
Best way to copy a subset of components from one entity to another?

### DISCUSSION THREAD

**Discussion by RaeASunshine** [genaray/Arch#51] [General]
So, I'm trying to copy some components from one entity to another entity, where the component structs are tagged with some attribute.

For context, I'm trying to copy components specific to rendering by tagging specific components with a `[ImportantToRendering]` attribute.  The idea is that I want to snapshot the state of the world at the end of a fixed simulation tick so that the renderer can interpolate from the previous state to the current state for variable rendering speed.  

I am able to grab, for a given entity, all of the components that should be mirrored over to the render state.  However, actually copying the ImportantToRendering tagged components on the entity that should be copied is eluding me.  

Any advice here on how to bulk copy some components from one entity to another?

**Comment by genaray** [comment]
Hey there ! :) 

Looks like you just found a missing feature. That was not on my screen at all. 
I will make an issue out of it and plan the feature.

There is a way to copy components (the chunks do this internally), but the way is a bit more manual work.

## If have access to the internals 

For now, you can try just that. You need to get the chunk of the two entities. From this chunk you get the component arrays and then you just have to copy the component manually. 

You must also make sure that both entities have the components to be copied.

For example, if you copy position and velocity from one entity to the other, both entities will need these components. If they don't exist on the entity yet, logically you have to add them first :) 

```csharp
var entityInfo = world.EntityInfo[entity.Id];
var otherEntityInfo = world.EntityInfo[otherEntity.Id];

var chunk = entity.GetChunk();
var otherChunk = otherEntity.GetChunk();

var firstArray = chunk.GetArray<...>();
var secondArray = otherChunk .GetArray<...>();

// Copy from item from first array into second array at the same position, entityInfo contains the slot of an entity inside its chunk. 
// Repeat for all components which should be copied. 
// Theres also chunk.GetArray(type) in case you cant use generics
```

## If you use the nuget

Then we have a problem... in this case you can't access the internals and since the copy feature doesn't exist yet you will have to use generics first.

```csharp
var toCopy = entity.Get<Component>();
if(otherEntity.Has<Component>()) otherEntity.Set(toCopy);
else otherEntity.Add(toCopy);

// repeat for other components
```

Im gonna work quickly on this to provide a clean copy mechanism :)

**Reply by RaeASunshine** [reply]
for what it's worth, I don't know the exact components at compile time that will be copied.  I query the entity's component types to get a list of them that are important for the renderer to copy:

```cs
var Types = e.GetComponentTypes().Where(t => t.Type.GetCustomAttribute<CopyToRenderWorldAtribute>() != null).ToArray();
```

I can pass this array into a world.Create() call just fine, I just can't copy the values from one entity to another this way, I need to know the component at compile time to do this (or fuss with some gross reflection code to do so).

**Reply by genaray** [reply]
I see... Hmm... I could probably add some overloads for `.Set, .Add, .Remove, .Get` which will accept `Type`s as params instead of generics. I could to that quickly i guess, lemme take a look at that.

**Reply by genaray** [reply]
Im almost done, currently writing tests to verify that it works correctly :) then im gonna push and release a new nugget.

**Reply by RaeASunshine** [reply]
awesome

**Reply by genaray** [reply]
> awesome

Thanks :) https://github.com/genaray/Arch/commit/fd9e375898d7e14f2be84ecf8144edd45bea38aa now included non generic alternatives which can be used to "copy" / set, add, remove, get data from entities by their runtime types :) 

Im gonna publish this real fast, the checks were all positive so i think it will work flawless.

**Reply by genaray** [reply]
Just released 1.1.2 featuring the non generic entity methods, some performance improvements and some small other changes. Nuget package is uploading, should be available within a few minutes :) If there some bugs i have not noticed, lemme know.

**Reply by genaray** [reply]
Looks like i forgot to include the remove,add methods to entityextensions, gonna prepare a hotfix real quick. 
Is now up and working :)
